File: go/dnse/api_account.go
package dnse

import (
	"fmt"
)

// GetAccounts retrieves all trading sub-accounts managed under the account corresponding to the API Key.
func (c *Client) GetAccounts(dryRun bool) (int, []byte, error) {
	return c.Request("GET", "/accounts", nil, nil, nil, dryRun)
}

// GetBalances retrieves asset balances of a trading sub-account.
func (c *Client) GetBalances(accountNo string, dryRun bool) (int, []byte, error) {
	return c.Request("GET", fmt.Sprintf("/accounts/%s/balances", accountNo), nil, nil, nil, dryRun)
}

// GetLoanPackages retrieves available loan package codes.
func (c *Client) GetLoanPackages(accountNo, marketType, symbol string, dryRun bool) (int, []byte, error) {
	query := map[string]string{
		"marketType": marketType,
	}
	if symbol != "" {
		query["symbol"] = symbol
	}
	return c.Request("GET", fmt.Sprintf("/accounts/%s/loan-packages", accountNo), query, nil, nil, dryRun)
}

// GetPositions retrieves current holding positions.
func (c *Client) GetPositions(accountNo, marketType string, dryRun bool) (int, []byte, error) {
	query := map[string]string{
		"marketType": marketType,
	}
	return c.Request("GET", fmt.Sprintf("/accounts/%s/positions", accountNo), query, nil, nil, dryRun)
}

// GetPositionByID retrieves detailed information of a specific position.
func (c *Client) GetPositionByID(marketType, positionID string, dryRun bool) (int, []byte, error) {
	query := map[string]string{
		"marketType": marketType,
	}
	return c.Request("GET", fmt.Sprintf("/accounts/positions/%s", positionID), query, nil, nil, dryRun)
}

// GetPPSE retrieves buying power and selling power before placing an order.
func (c *Client) GetPPSE(accountNo, marketType, symbol string, price float64, loanPackageID string, dryRun bool) (int, []byte, error) {
	query := map[string]string{
		"marketType":    marketType,
		"symbol":        symbol,
		"price":         fmt.Sprintf("%v", price),
		"loanPackageId": loanPackageID,
	}
	return c.Request("GET", fmt.Sprintf("/accounts/%s/ppse", accountNo), query, nil, nil, dryRun)
}

// ClosePosition closes an existing position (by ID).
func (c *Client) ClosePosition(positionID, marketType, tradingToken string, dryRun bool) (int, []byte, error) {
	headers := map[string]string{
		"trading-token": tradingToken,
	}
	query := map[string]string{
		"marketType": marketType,
	}
	return c.Request("POST", fmt.Sprintf("/accounts/positions/%s/close", positionID), query, nil, headers, dryRun)
}

File: rust/src/rest/api_account.rs
use super::client::DnseClient;
use reqwest::Method;
use std::collections::HashMap;

impl DnseClient {
    pub async fn get_accounts(&self, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        self.request(Method::GET, "/accounts", None, None, None, dry_run).await
    }

    pub async fn get_balances(&self, account_no: &str, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        self.request(Method::GET, &format!("/accounts/{}/balances", account_no), None, None, None, dry_run).await
    }

    pub async fn get_loan_packages(&self, account_no: &str, market_type: &str, symbol: Option<&str>, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        let mut query = HashMap::new();
        query.insert("marketType", market_type.to_string());
        if let Some(s) = symbol {
            query.insert("symbol", s.to_string());
        }
        self.request(Method::GET, &format!("/accounts/{}/loan-packages", account_no), Some(&query), None, None, dry_run).await
    }

    pub async fn get_positions(&self, account_no: &str, market_type: &str, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        let mut query = HashMap::new();
        query.insert("marketType", market_type.to_string());
        self.request(Method::GET, &format!("/accounts/{}/positions", account_no), Some(&query), None, None, dry_run).await
    }

    pub async fn get_position_by_id(&self, market_type: &str, position_id: &str, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        let mut query = HashMap::new();
        query.insert("marketType", market_type.to_string());
        self.request(Method::GET, &format!("/accounts/positions/{}", position_id), Some(&query), None, None, dry_run).await
    }

    pub async fn get_ppse(&self, account_no: &str, market_type: &str, symbol: &str, price: f64, loan_package_id: &str, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        let mut query = HashMap::new();
        query.insert("marketType", market_type.to_string());
        query.insert("symbol", symbol.to_string());
        query.insert("price", price.to_string());
        query.insert("loanPackageId", loan_package_id.to_string());
        self.request(Method::GET, &format!("/accounts/{}/ppse", account_no), Some(&query), None, None, dry_run).await
    }

    pub async fn close_position(&self, position_id: &str, market_type: &str, trading_token: &str, dry_run: bool) -> Result<(u16, Vec<u8>), Box<dyn std::error::Error>> {
        let mut query = HashMap::new();
        query.insert("marketType", market_type.to_string());

        let mut headers = HashMap::new();
        headers.insert("trading-token", trading_token.to_string());

        self.request(Method::POST, &format!("/accounts/positions/{}/close", position_id), Some(&query), None, Some(&headers), dry_run).await
    }
}
